batched/dense/unit_test/Test_Batched_SerialGbtrf.hpp

```diff
@@ -0,0 +1,519 @@
+  banded_to_full<View3DType, View3DType>(AB2, A2, kl, ku);
+
+  // Copy upper triangular components to U
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Upper, KokkosBatched::Diag::NonUnit>(A0, U0);
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Upper, KokkosBatched::Diag::NonUnit>(A1, U1);
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Upper, KokkosBatched::Diag::NonUnit>(A2, U2);
+
+  // Extract L
+  // Apply pivot at host
+  auto h_ipiv0 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), ipiv0);
+  auto h_ipiv1 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), ipiv1);
+  auto h_ipiv2 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), ipiv2);
+  auto h_A0    = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A0);
+  auto h_A1    = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A1);
+  Kokkos::deep_copy(h_A2, A2);
+  for (int ib = 0; ib < Nb; ib++) {
+    for (int j = 0; j < BlkSize - 1; j++) {
+      for (int i = j + 1; i < BlkSize - 1; i++) {
+        Kokkos::kokkos_swap(h_A0(ib, i, j), h_A0(ib, h_ipiv0(ib, i), j));
+      }
+    }
+
+    for (int j = 0; j < BlkSize - 2; j++) {
+      for (int i = j + 1; i < BlkSize - 1; i++) {
+        Kokkos::kokkos_swap(h_A1(ib, i, j), h_A1(ib, h_ipiv1(ib, i), j));
+      }
+    }
+
+    for (int j = 0; j < BlkSize; j++) {
+      for (int i = j + 1; i < BlkSize - 1; i++) {
+        Kokkos::kokkos_swap(h_A2(ib, i, j), h_A2(ib, h_ipiv2(ib, i), j));
+      }
+    }
+  }
+  Kokkos::deep_copy(A0, h_A0);
+  Kokkos::deep_copy(A1, h_A1);
+  Kokkos::deep_copy(A2, h_A2);
+
+  // Copy non-diagonal lower triangular components to NL
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Lower, KokkosBatched::Diag::NonUnit>(A0, NL0,
+                                                                                                             -1);
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Lower, KokkosBatched::Diag::NonUnit>(A1, NL1,
+                                                                                                             -1);
+  create_triangular_matrix<View3DType, View3DType, KokkosBatched::Uplo::Lower, KokkosBatched::Diag::NonUnit>(A2, NL2,
+                                                                                                             -1);
+
+  // Check if U, NL and ipiv have expected values
+  auto h_U0  = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), U0);
+  auto h_U1  = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), U1);
+  auto h_U2  = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), U2);
+  auto h_NL0 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), NL0);
+  auto h_NL1 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), NL1);
+  auto h_NL2 = Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), NL2);
+
+  RealType eps = 1.0e1 * ats::epsilon();
```

Comment:
This looks like an arbitrary number that happens to work... how about doing an error analysis to compute the number of round off operations performed in gbtrf?

Reply:
Could you please detail this point?
The tolerance is numerical precision of fp32 or fp64 multiplied by 10.

Reply:
For example when you perform a gemv operation: 

> `y = beta * y + alpha * A * x`

for each value `y(i)` you have performed `numCols` multiplications and `numCols - 1` additions to compute `A * x`, then there is two more multiplications for `alpha` and `beta` and one more addition between `beta * y` and `alpha * A * x` so in total that's `numCols + 2` multiplications and `numCols` additions. So a check might look like this

> `tol = (2 * numCols + 2) * maxVal * Kokkos::ArithTraits<Scalar>::eps()`
> `Kokkos::abs(y(i) - y_ref(i)) < tol`

the maxVal is the maximum value an input can take as catastrophic cancelation could happen

Reply:
I understand your point. 
The error analysis would be critical, when it comes to fp16.

Can I start from simpler cases like `gemv` and `gemm`.